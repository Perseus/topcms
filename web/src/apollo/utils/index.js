export const commonQueryFields = () => ( `
  code
  message
  success
  errors
` );
