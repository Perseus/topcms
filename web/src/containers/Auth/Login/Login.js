import { mapState, mapActions } from 'vuex';
import * as ActionTypes from '../../../store/action-types';
import RouteNames from '../../../config/RouteNames';

import errorHandlerMixin from '../../../mixins/errorHandler';
import Button from '../../../components/Button/Button.vue';
import TextInput from '../../../components/TextInput/TextInput.vue';
import FormError from '../../../components/FormError.vue';

const Login = {

  data() {
    return {
      username: '',
      password: '',
      errors: [],
    };
  },

  components: {
    Button,
    TextInput,
    'form-error': FormError
  },

  mixins: [ errorHandlerMixin ],

  methods: {
    ...getActionDispatchers(),
  },
  computed: {},

  watch: {

  }


};

function getActionDispatchers() {
  return mapActions( {
    loginUser: `user/${ActionTypes.loginUser}`,
    changeRoute: `router/${ActionTypes.changeRoute}`,
  } );
}

export default Login;
