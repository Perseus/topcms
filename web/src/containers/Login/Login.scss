.login-form {
  display: flex;
  justify-content: center;
  align-items: center;
}

.cta {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
}

.sign-up-cta {
  font-size: 12px;
}

.forgot-password {
  font-size: 12px;
}

.login-btn {
  margin-top: 15px;
}

.small-action-group {
  display: flex;
  align-items: center;
  width: 100%;
  justify-content: space-between;
}
