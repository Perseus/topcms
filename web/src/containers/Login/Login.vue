<template>
  <section class="container login-form">
    <div class="card">
      <div class="card-content">
        <ValidationObserver v-slot="{ handleSubmit }" ref="loginForm">
          <form @submit.prevent="handleSubmit(onUserLogin)">
            <TInput label="Username" rules="required" v-model="username" name="Username"/>
            <TInput label="Password" type="password" rules="required|min:5" v-model="password" name="Password"/>

            <div class="cta">
              <div class="small-action-group">
                <a
                  href="#"
                  class="forgot-password"
                  @click.prevent="redirectToForgot"
                >Forgot Password?</a>
                <a href="#" class="sign-up-cta" @click.prevent="redirectToSignup">Sign Up</a>
              </div>
              <b-button
                type="is-primary"
                native-type="submit"
                class="login-btn"
                :loading="isUserLoggingIn"
              >
                Login
              </b-button>
            </div>
          </form>
        </ValidationObserver>
      </div>
    </div>
  </section>
</template>


<script src="./Login.js"></script>
<style src="./Login.scss" lang="scss" scoped></style>