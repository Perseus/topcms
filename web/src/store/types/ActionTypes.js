const ActionTypes = {
  bootstrapApplication: 'bootstrapApplication',
  retrieveLandingPageInformation: 'retrieveLandingPageInformation',
  retrieveStaffOnlineStatus: 'retrieveStaffOnlineStatus',
  connectToSocketServer: 'connectToSocketServer',

  registerUser: 'registerUser',
  loginUser: 'loginUser',
  retrieveUser: 'retrieveUser',
  logoutUser: 'logoutUser',
  updateUser: 'updateUser',

  setInitialRoute: 'setInitialRoute',
  changeRoute: 'changeRoute',
  changeRouteWithoutPlugin: 'changeRouteWithoutPlugin',
  updateRequestsInProgress: 'updateRequestsInProgress',
  toggleModal: 'toggleModal',

  getServerStats: 'getServerStats',
  getSiteAuthors: 'getSiteAuthors',
  getSiteNewsArticles: 'getSiteNewsArticles',
  getSiteDownloads: 'getSiteDownloads',
  getSitePolls: 'getSitePolls',
  getAllSiteInfo: 'getAllSiteInfo',
  getSiteNewsFeed: 'getSiteNewsFeed',
  addItemToNewsFeed: 'addItemToNewsFeed',

  createSiteAuthor: 'createSiteAuthor',
  updateSiteAuthor: 'updateSiteAuthor',
  deleteSiteAuthor: 'deleteSiteAuthor',
  createSiteDownload: 'createSiteDownload',
  updateSiteDownload: 'updateSiteDownload',
  deleteSiteDownload: 'deleteSiteDownload',
  createSiteNews: 'createSiteNews',
  updateSiteNews: 'updateSiteNews',
  deleteSiteNews: 'deleteSiteNews',

  getSiteDownload: 'getSiteDownload',
  getSiteNewsArticle: 'getSiteNewsArticle',
  getSiteAuthor: 'getSiteAuthor',

  fetchServerRates: 'fetchServerRates',
  updateServerRates: 'updateServerRates',

  retrievePlayerRanking: 'retrievePlayerRanking',
  retrieveGuildRanking: 'retrieveGuildRanking',

  retrieveFilteredAccounts: 'retrieveFilteredAccounts',
  retrieveFilteredCharacters: 'retrieveFilteredCharacters',
  toggleBanForUser: 'toggleBanForUser',
  retrieveAccountData: 'retrieveAccountData',
  adminUpdateUserEmail: 'adminUpdateUserEmail',
  adminUpdateUser: 'adminUpdateUser',
  retrieveCharacter: 'retrieveCharacter',
  generateItemInfoCache: 'generateItemInfoCache',
  uploadItemInfo: 'uploadItemInfo',
  resetUserSecurityCode: 'resetUserSecurityCode',

  retrieveMallCategories: 'retrieveMallCategories',
  createMallCategory: 'createMallCategory',
  editMallCategory: 'editMallCategory',
  deleteMallCategory: 'deleteMallCategory',

  retrieveMallItems: 'retrieveMallItems',
  createMallItem: 'createMallItem',
  editMallItem: 'editMallItem',
  deleteMallItem: 'deleteMallItem',

  addMallPoints: 'addMallPoints',
  purchaseMallItem: 'purchaseMallItem',
};

export default ActionTypes;
