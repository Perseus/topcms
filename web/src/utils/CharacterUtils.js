import { CharacterConfig } from '../config/CharacterConfig';

export function hasValidJobIcon( character ) {
  const icon = character.icon.toLowerCase();
  const job = character.job.toLowerCase();

  if ( !icon || !job ) {
    return false;
  }

  if ( !CharacterConfig.VALID_JOBS[ icon ] ) {
    return false;
  }

  if ( !CharacterConfig.VALID_JOBS[ icon ].includes( job ) ) {
    return false;
  }

  return true;
}
