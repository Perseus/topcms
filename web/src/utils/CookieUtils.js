export function getCookie( name ) {
  const cookie = {};
  document.cookie.split( ';' ).forEach( ( el ) => {
    const [ k, v ] = el.split( '=' );
    cookie[ k.trim() ] = v;
  } );
  return cookie[ name ];
}
