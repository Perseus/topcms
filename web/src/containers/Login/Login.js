import { mapActions, mapGetters } from 'vuex';
import { BButton } from 'buefy/dist/components/button';

import ActionTypes from '../../store/types/ActionTypes';
import RouteNames from '../../config/RouteNames';
import TInput from '../../components/ValidationInputs/TInput.vue';


const Login = {
  name: 't-login',
  data() {
    return {
      username: '',
      password: '',
    };
  },

  components: {
    TInput,
    'b-button': BButton,
  },

  computed: {
    ...getStateGetters(),
  },


  methods: {
    ...getActionDispatchers(),
    onUserLogin() {
      this.loginUser( {
        username: this.username,
        password: this.password
      } );
    },
    redirectToSignup() {
      this.changeRoute( {
        name: RouteNames.AUTH.REGISTER
      } );
    },
    redirectToForgot() {

    }
  }
};

function getActionDispatchers() {
  return mapActions( {
    loginUser: ActionTypes.loginUser,
    changeRoute: ActionTypes.changeRoute
  } );
}

function getStateGetters() {
  return mapGetters( {
    isUserLoggingIn: 'isUserLoggingIn',
  } );
}

export default Login;
