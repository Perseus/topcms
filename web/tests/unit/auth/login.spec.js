import faker from 'faker';

import Vuex from 'vuex';
import Login from '@containers/Login/Login.vue';
import userModule from '@store/modules/user';
import request from '@services/GraphQLRequest';
import ActionTypes from '@store/types/ActionTypes';
import { mountComponentWithLocalVue, flushPromises } from '../utils/localVue';

jest.mock( '@services/GraphQLRequest' );

request.mutation.mockImplementation( ( _, variables ) => Promise.resolve( {
  loginUser: {
    code: 'OK',
    message: null,
    success: true,
    errors: null,
    data: {
      name: variables.name, account_details: { access_levels: [ 2, 0, 1 ], __typename: 'Account' }, __typename: 'User'
    },
    __typename: 'UserResponse'
  }
} ) );

const loginUser = jest.fn();
const mountLoginComponent = ( mountOptions, storeOptions ) => mountComponentWithLocalVue( Login, mountOptions, {
  getters: {
    isUserLoggingIn: () => false,
  },
  actions: {
    loginUser
  },
  ...storeOptions
} );

describe( 'Login Component', () => {
  it( 'renders correctly', () => {
    const wrapper = mountLoginComponent();

    expect( wrapper ).toMatchSnapshot();
  } );

  it( 'displays validation errors if the form is submitted without any data', async() => {
    const wrapper = mountLoginComponent();

    const loginButton = wrapper.find( '[data-test="login-button"]' );
    await loginButton.trigger( 'submit' );

    await flushPromises();

    const errorEls = wrapper.findAll( '.help' );
    errorEls.wrappers.forEach( wrap => expect( wrap.text() ).not.toHaveLength( 0 ) );
  } );

  it( 'triggers the loginUser function on valid form submission', async() => {
    const wrapper = mountLoginComponent( {
      data: () => ( {
        username: faker.name.findName(),
        password: faker.internet.password(),
      } )
    } );

    const loginButton = wrapper.find( '[data-test="login-button"]' );
    await loginButton.trigger( 'submit' );

    await flushPromises();

    expect( loginUser ).toHaveBeenCalled();
  } );

  it( 'loginUser action triggers graphql request', async() => {
    const store = new Vuex.Store( {
      modules: {
        userModule
      }
    } );

    const data = {
      username: faker.name.findName(),
      password: faker.internet.password()
    };

    await store.dispatch( ActionTypes.loginUser, data );

    expect( request.mutation ).toHaveBeenCalled();
  } );

  it( 'loginUser action commits SIGNIN_COMPLETE and dispatches changeRoute on response', async() => {
    const SIGNIN_COMPLETE = jest.fn();
    const changeRoute = jest.fn();

    const store = new Vuex.Store( {
      modules: {
        userModule: {
          ...userModule,
          mutations: {
            SIGNIN_COMPLETE,
          },
          actions: Object.assign( userModule.actions, {
            changeRoute
          } )
        }
      }
    } );

    const data = {
      username: faker.name.findName(),
      password: faker.internet.password()
    };

    await store.dispatch( ActionTypes.loginUser, data );

    expect( SIGNIN_COMPLETE ).toHaveBeenCalled();
    expect( changeRoute ).toHaveBeenCalled();
  } );
} );
