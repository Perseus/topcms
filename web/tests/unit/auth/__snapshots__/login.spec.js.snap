// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component renders correctly 1`] = `
<section class="container login-form">
  <div class="card">
    <div class="card-content"><span><form><span label="Username" name="Username"><div class="field" name="Username"><label class="label">Username</label> <div class="control"><input type="text" autocomplete="off" data-test="Username" label="Username" name="Username" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span> <span label="Password" type="password" name="Password"><div class="field" name="Password"><label class="label">Password</label> <div class="control"><input type="password" autocomplete="off" data-test="Password" label="Password" name="Password" class="input"> <!----> <!----> <!----></div> <p class="help"></p></div></span>
      <div class="cta">
        <div class="small-action-group"><a href="#" class="forgot-password">Forgot Password?</a> <a href="#" class="sign-up-cta">Sign Up</a></div> <button type="submit" data-test="login-button" class="button login-btn is-primary">
          <!----> <span>
              Login
            </span>
          <!----></button>
      </div>
      </form>
      </span>
    </div>
  </div>
</section>
`;
